File: frontend/src/api/portfolio.ts
/**
 * 自选组合 API 服务
 * 封装组合和标的管理相关请求
 */

import apiClient from '@/api/client';
import type { Portfolio, WatchlistItem, MessageResponse } from '@/types';

/** 添加标的请求 */
interface AddItemRequest {
  symbol: string;
  name: string;
  market: string;
}

const portfolioApi = {
  /** 获取用户所有组合 */
  getPortfolios: () =>
    apiClient.get<Portfolio[]>('/api/portfolios'),

  /** 创建新组合 */
  createPortfolio: (name: string) =>
    apiClient.post<Portfolio>('/api/portfolios', { name }),

  /** 更新组合（重命名） */
  updatePortfolio: (id: number, name: string) =>
    apiClient.put<Portfolio>(`/api/portfolios/${id}`, { name }),

  /** 删除组合 */
  deletePortfolio: (id: number) =>
    apiClient.delete<MessageResponse>(`/api/portfolios/${id}`),

  /** 调整组合排序 */
  reorderPortfolios: (ids: number[]) =>
    apiClient.put<MessageResponse>('/api/portfolios/reorder', { ids }),

  /** 获取组合内标的（含行情） */
  getItems: (portfolioId: number) =>
    apiClient.get<WatchlistItem[]>(`/api/portfolios/${portfolioId}/items`),

  /** 添加标的到组合 */
  addItem: (portfolioId: number, data: AddItemRequest) =>
    apiClient.post<WatchlistItem>(`/api/portfolios/${portfolioId}/items`, data),

  /** 移除标的 */
  removeItem: (portfolioId: number, itemId: number) =>
    apiClient.delete<MessageResponse>(`/api/portfolios/${portfolioId}/items/${itemId}`),

  /** 调整标的排序 */
  reorderItems: (portfolioId: number, ids: number[]) =>
    apiClient.put<MessageResponse>(`/api/portfolios/${portfolioId}/items/reorder`, { ids }),
};

export default portfolioApi;

File: backend/app/routers/portfolio.py
"""
自选组合路由
处理组合的增删改查和排序
"""

from __future__ import annotations

from fastapi import APIRouter, Depends

from app.core.deps import get_current_active_user
from app.core.memory_store import UserData
from app.schemas.common import MessageResponse
from app.schemas.portfolio import (
    CreatePortfolioRequest,
    PortfolioResponse,
    ReorderPortfoliosRequest,
    UpdatePortfolioRequest,
)
from app.services.portfolio_service import PortfolioService

router = APIRouter(prefix="/api/portfolios", tags=["自选组合"])

portfolio_service = PortfolioService()


@router.get("", response_model=list[PortfolioResponse])
async def get_portfolios(current_user: UserData = Depends(get_current_active_user)):
    """获取用户所有自选组合"""
    return await portfolio_service.get_portfolios(current_user.id)


@router.post("", response_model=PortfolioResponse, status_code=201)
async def create_portfolio(
    data: CreatePortfolioRequest,
    current_user: UserData = Depends(get_current_active_user),
):
    """创建新自选组合"""
    return await portfolio_service.create_portfolio(current_user.id, data)


@router.put("/reorder", response_model=MessageResponse)
async def reorder_portfolios(
    data: ReorderPortfoliosRequest,
    current_user: UserData = Depends(get_current_active_user),
):
    """调整组合排序"""
    await portfolio_service.reorder_portfolios(current_user.id, data.ids)
    return MessageResponse(message="排序更新成功")


@router.put("/{portfolio_id}", response_model=PortfolioResponse)
async def update_portfolio(
    portfolio_id: int,
    data: UpdatePortfolioRequest,
    current_user: UserData = Depends(get_current_active_user),
):
    """更新组合信息（重命名）"""
    return await portfolio_service.update_portfolio(portfolio_id, current_user.id, data)


@router.delete("/{portfolio_id}", response_model=MessageResponse)
async def delete_portfolio(
    portfolio_id: int,
    current_user: UserData = Depends(get_current_active_user),
):
    """删除组合（默认组合不可删除）"""
    await portfolio_service.delete_portfolio(portfolio_id, current_user.id)
    return MessageResponse(message="组合删除成功")
